cou/steps/analyze.py

```diff
@@ -71,3 +81,16 @@ def __str__(self) -> str:
         :rtype: str
         """
         return os.linesep.join([str(app) for app in self.apps])
+
+    @property
+    def current_cloud_os_release(self) -> Optional[OpenStackRelease]:
```

Comment:
at the moment we only have a single type of Application class, and use `current_os_release` being `None` to mean several things: "not an OpenStack app", "an OpenStack app undergoing a per-unit upgrade", "an OpenStack app in a weird state". We then calculate the overall cloud version based only on the apps that have a consistent `current_os_release`. This may be ok in this WIP phase, but would lead to incorrect behavior later on (unless modified): imagine an upgrade that is almost complete, save for one last app running into an issue leading to mismatched versions. You re-run cou and suddenly get offered to start the next upgrade transition.

I think for now we should either raise an exception here or return None whenever there is at least one app with `current_os_release == None`.

Later on, when we have more sophisticated classes (e.g. `NotAnOpenStackApplication`, `NovaCompute`, etc) we can be smarter about whether it's ok or not to ignore an app.

Reply:
Now we are like this:
- `Non - OpenStack app` is skipped because there is no source or openstack-origin and a debug log was addded
- OpenStack app in a weird state raises exception. This is currently including the situation of upgrading per-unit
- skipping apps with `current_os_release == None`. This is currently happening on subordinate charms (until we implement a proper class) because there are no units on ApplicationStatus and also when we are not able to identify the OpenStack release because the workload version is unknown. On both cases I'm using warning logs.